Added Mutators to support changing object on read/write

lordnull opened the pull request:
This is to support https://github.com/basho/riak_repl/issues/223 .

This implements a 'mutator' interface. A mutator callback module can alter an object to be stored and after it has been retrieved from a storage backend.

Note this has been recently rebased and force-pushed (sept 26) to account for changes from develop.

lordnull committed:
Step one in reduced replications: a hook to mutate.

To achieve reduced replications, a mutation of the object on put and get
needs to be applied, and needs to be loosely coupled to the systems that
will be doing the mutating (repl). This module provides a place to
register and run mutators on puts and gets.

lordnull committed:
Refactored to use the new metadata system.

Rather than try to manage the list itself, it leans heavily on the
metadata system to manage diferences and storage of the data.

lordnull committed:
Added test for when no mutators ever registered.

lordnull committed:
Made get mutation dependant on a put mutation.

This means making mutators is safer as they are expected to be
symetrical. A get mutator is only called if the object in question was
put through the corresponding put mutator.

lordnull committed:
Gets mutators no longer get bucket properties.

The get in kv_vnode does not have the bucket properties available,
so to avoid additional slow-down, the get's no longer require bucket
properties.

lordnull committed:
Hooked meta data into get/put path and tests.

Reliable tests remain reliable.

lordnull committed:
Fixed mutator mutation order.

Assuming the mutator module is correct*, mutations on get are applied in
the reverse order from put.

A mutator module is "correct" if the get is the reverse function of a put.
All the data a mutator needs to reverse a given mutation must be put into
the object to be able to do this.

lordnull committed:
added priorities to mutators.

Currently defualt is 0. There's no restriction on what a priority can
be since all erlang terms can be compared in < and >.

lordnull committed:
Fixed typo

lordnull committed:
Refactored / changed mutator to be more resilient.

Callback modules can now be set to only worry about a given meta/value
pair at a time, but also gets full object. Put can expose some meta data
to the putting client that will be different from the stored data.

lordnull committed:
Fixed vnode call to mutator system.

lordnull committed:
refactored get mutation again.

It become needed for a mutator to know which specific meta/value pair it
was working on. Rather than extend the function, I've opted to just
give the callback the object directly.

lordnull committed:
Get mutators can now (legitimately) return notfound

lordnull committed:
Allowed mutators to use notfound as a response on get.

lordnull committed:
Much documentation for mutation.

lordnull committed:
Fixed resolver for mutators.

Vagabond commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,210 @@
+%% <code><b>mutate_put(MetaData, Value, ExposedMeta,
+%% FullObject, BucketProperties) -> Result</b></code>
+%%
+%% Types:
```
I think edoc can generate all this from dialyzer -type and -callback specs now.

lordnull replied:
Looking into this (and trying it as well), edoc lists the functions but doesn't give type information.

Vagabond replied:
I think if you did like -type metadata() :: dict(). and then used metadata() it'd show them better?

lordnull replied:
I think we need to define better. I'd love to be able to have it automatically use the correct types, but that's a ways out. Adding links to types is a next step, but would it be better inline (in the code element) or in the expanded area?

Vagabond commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,210 @@
+
+-module(riak_kv_mutator).
+
+-include_lib("eunit/include/eunit.hrl").
```
Shouldn't this be -ifsef(TEST)?

Vagabond commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,210 @@
+%% can only be registered once.
+-spec register(Module :: atom(), Priority :: term()) -> 'ok'.
+register(Module, Priority) ->
+    Modifier = fun
```
It'd be nice to comment what resolution strategy this is trying to implement.

Vagabond commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,210 @@
+-spec get() -> [atom()].
+get() ->
+    Resolver = fun
+        ('$deleted', '$delete') ->
```
Is the second atom here a typo?

lordnull committed:
Removed eunit test hrl inclusion

Hold over from when the eunit tests were at the bottom of the module.
As they are not, this should have been removed.

lordnull committed:
Added comment about conflict resultion on mutators

lordnull committed:
Fixed typo.

Vagabond commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,209 @@
+mutate_put(Object, BucketProps) ->
+    Contents = riak_object:get_contents(Object),
+    {ok, Modules} = ?MODULE:get(),
+    MetasValuesRevealeds = lists:map(fun({InMeta, InValue}) ->
```
So for each sibling in the object, we apply all the mutators, then replace the contents of the object with the result?

Vagabond replied:
Really, more comments here would be nice.

lordnull replied:
Hopefully 5934e3b will address this particular issue.

lordnull committed:
Added comments and documentation

More than what, but some why too!

Vagabond commented:
The tests fail for me, presumably because of changes to cluster metadata:

```
=ERROR REPORT==== 24-Sep-2013::15:38:43 ===
** Generic server riak_core_metadata_manager terminating
** Last message in was {put,{{riak_kv,mutators},list},
                            [],#Fun<riak_kv_mutator.0.123063518>}
** When Server state == {state,nonode@nohost,"test_data",876565}
** Reason for termination ==
** {noproc,{gen_server,call,
                       [riak_core_metadata_hashtree,
                        {insert,{{riak_kv,mutators},list},
                                <<0,170,139,16,5,28,164,71,218,221,64,20,238,
                                  83,39,191,48,160,195,137>>,
                                false},
                        infinity]}}
```

Vagabond commented:
I'm mostly OK with this change, although I think @jtuple was also going to look at it.

lordnull committed:
Added capability declaration

lordnull committed:
Fixed setup/teardown to reflect metadata requirements

Tests should pass without warnings now.

lordnull commented:
Eunit tests should pass now.

Also, there is now a riak_test pull request that enhances the tests further; it attempts to stress the interaction between aae and the replication mutator: https://github.com/basho/riak_test/pull/392

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,222 @@
+%% data storage backend is not meant to be the object given. An
+%% example would be storing only meta data for an object on a remote
+%% cluster. This module is an interface to register mutators that will
+%% can be run.
```
I can haz this sentence fixed next time you're in the file?

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,222 @@
+    ?MODULE:register(Module, ?DEFAULT_PRIORITY).
+
+%% @doc Register a module as a mutator with the given priority. Modules with
+%% equal priority are done in sort sort (alphabetical) order. A module
```
sort sort == sort ^ 2 ?

lordnull committed:
Fixed grammatical errors in docs.

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,222 @@
+    % {Meta, Value, Exposed} values, the get callback gets away with just
+    % giving the object? This is to avoid complicated interaction with
+    % notfound return.
+    case Mutator:mutate_get(Object) of
```
Notice that this makes it possible to invalidate all the data in the cluster if a mutator stored in object metadata is removed from the code. Users of this feature will have to be careful about some downgrade scenarios where data is written with a mutator in the list than then disappears on downgrade.

lordnull replied:
Do we just leave this to documentation, or should there be some insurance put in the code here?

engelsanchez replied:
A note in the documentation would be enough for now.  It's good to flag the worse case scenarios.

engelsanchez commented on src/riak_kv_app.erl:
```diff
@@ -187,6 +187,9 @@ start(_Type, _StartArgs) ->
                                           disabled),
 
+            % get/put mutator is supported, but not all nodes may have it
+            riak_core_capability:register({riak_kv, mutators}, [true, false], false),
```
This capability is not being queried anywhere that I can see.

lordnull replied:
It is used in repl before attempting to install the reduced repl mutator: https://github.com/basho/riak_repl/blob/develop/src/riak_repl_sup.erl#L20

lordnull committed:
Added warning about removing mutator code.

Added a similar warning in 4 places concerning the same thing. If an object
is stored using a given mutator, that same mutator is used when the object
is retreived. If that code isn't available or has changed in a backwards
incompatible way, one can expect to have a bad time.

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,240 @@
+    end,
+    ModulesAndPriors = riak_core_metadata:get({riak_kv, mutators}, list, [{default, []}, {resolver, Resolver}]),
+    Flipped = [{P, M} || {M, P} <- ModulesAndPriors],
+    Sorted = lists:sort(Flipped),
```
A sort in the critical path could become an issue with multiple mutators. I know we are not planning to add those soon, but it seems that the resolution and merge functions could take care of this directly and store values in sorted order, no?

engelsanchez commented:
While writing my own test mutators I was surprised by how they can be called twice on the same object (once in the local put, and the result of that is passed to the other vnodes in the preflist).  Unless the application of the mutator chain is completely idempotent, it will be very hard to reason about.  I had tried to do a simple "make value a list of two values" and a "make a list a tuple" mutator and put them in a chain, but the results end up just crazy.  The first one went from a value of 1 to {1, 1}, whereas the replicas ended up with {{1,1}, {1,1}}.  I'm leaving this here as a note of the headache that it could be to use this feature in a truly generic fashion.

engelsanchez commented:
I'm going to +1 this PR with the following caveats:
- We need to make sure we revisit and review performance before release. Part of my concerns will probably be addressed in the cluster metadata code by removing gen_server calls. The sort on get should be targeted for removal if resolution can create sorted lists without issue.
- I have my doubts that the code will survive as is when a second use case arrives. Of course, it's impossible to foresee what that may be.

The code as is should be enough to support the Repl use case.

lordnull committed:
Only doing a sort on mutation in worst case

Instead of storing a list of mutators as {mutator, priority}, then doing
a sort on every retrieval of the list of mutators, the mutators get stored
{Priority, Module}, so the list just gets the priority stripped off (faster
then a sort). This means there's a sort done in the following cases:
 - registering or unregistering a module. There's a check if the module
existed, and it's overwrittne if it did.
 - there are sibling values in metadata. A functionaly equal merge is
done on those values. Hopefully that is a short-lived situation.

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,247 @@
+        {P1, _} when P1 < Priority ->
+            Acc;
+        Else ->
+            ordsets:all_element(Mutator, ordsets:del_element(Else))
```
I haven't even run it, but you surely meant add_element here, not all_element.

lordnull replied:
D'oh!

lordnull committed:
Fixed typo

engelsanchez commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,247 @@
+-spec get() -> [atom()].
+get() ->
+    Resolver = fun
+        ('$deleted', '$deleted') ->
```
A note for @jrwest: it seems that we should have the metadata tombstone in a public header or is there a way for resolvers not to use this atom directly?

jrwest replied:
hrm thats a good point. we certainly want to expose tombstones to resolvers (or at least have the option to) but leaking this is a bit nasty. a macro would probably be quickest, but i wonder if there is a better way...

jrwest replied:
for this code specifically, does it even call `riak_core_metadata:delete`? If not, don't really need to have these here.

engelsanchez replied:
Well no, but I think it's worth handling anyway for our convenience at the very least. I used delete while testing it when the format was changed, for example.

engelsanchez commented:
I've looked over the sort change and tested manually. After the one issue I found was resolved, it looks good. Re +1 on this.

russelldb commented on src/riak_kv_mutator.erl:
```diff
@@ -0,0 +1,247 @@
+%% a crash.
+-spec mutate_get(Object :: riak_object:riak_object()) -> riak_object:riak_object().
+mutate_get(Object) ->
+    [Meta | _] = riak_object:get_metadatas(Object),
```
Is this right? What about siblings with no mutators in their metadata being higher up the list?
